### DISCUSSION TITLE
Adding an extra payment gateway integration

### DISCUSSION THREAD

**Discussion by jerome2710** [solspace/craft-freeform#420] [Questions & Answers]
Freeform ships with Stripe as the only payment gateway integration. While Europe often uses other PSP's such as [Mollie](https://www.mollie.com), we need to implement this as a payment integration instead of Stripe for a few of our clients. Is there any documentation available on how to properly implement another gateway? The `PaymentGatewaysService::EVENT_FETCH_TYPES` event allows to register another type, but it would be helpful to have some guidance implementing the methods.

On a side note; non-credit card payment methods often use off-site payment. The user completes the payment in a portal of their bank and return to the website afterwards. This is different than the PaymentIntents of Stripe, which will be a challenge as well.

**Comment by jerome2710** [comment]
I also have the feeling that Stripe is hard-wired into the Freeform-payments, not allowing other integrations, despite being the intention.

The `Form::EVENT_AFTER_VALIDATE` will trigger `StripeService::preProcessPayment()`, even if another gateway has been chosen in the form:
- https://github.com/solspace/craft-freeform/blob/master/packages/plugin/src/Bundles/Payments/Stripe/StripePaymentsBundle.php#L19
- https://github.com/solspace/craft-freeform/blob/master/packages/plugin/src/Services/Pro/Payments/StripeService.php#L43

**Comment by kjmartens** [comment]
Hi @jerome2710,

Freeform Payments was designed to allow for more payment gateways in the future, but it's not something that has been fully implemented yet, and you're right that several parts of the Stripe Payments feature are baked in. When we expand the gateway offerings in the future, this is something that will definitely be addressed. 🙂 

In the meantime, it currently isn't possible to support other payment gateways.

**Comment by jerome2710** [comment]
Hi @kjmartens,

That's unfortunate. The foundation has been laid, but it's not usable yet. Any idea when this will be finalized?

We need to find an alternative solution in the meantime. I feel like it's hard to extend the Javascript builder and add a custom payment component to the right sidebar in Craft. It might be easier to create a custom field and store the payment-settings there, without actually showing any inputs to the user. Is there any documentation on how to add a custom field with multiple settings? The [Field Events](https://docs.solspace.com/craft/freeform/v3/developer/field-events.html) page does not document `FieldsService::EVENT_FETCH_TYPES`.

**Reply by jerome2710** [reply]
Additionally for example, how do I render the CP template for adding a field of my custom field type? [templates/fields/edit.html](https://github.com/solspace/craft-freeform/blob/master/packages/plugin/src/templates/fields/edit.html#L108) assumes the field type type is one of Freeform's predefined, not allowing to add custom field type templates.

**Reply by kjmartens** [reply]
Unfortunately there currently isn't any way to add your own custom field types in Freeform. This is something else we are considering adding support for in the future, but either of these features could be many months away.
